SMTP issues during github action pytest

________________________ test_user_registration_success ________________________
async_client = <httpx.AsyncClient object at 0x7f5bd333d9d0>
    @pytest.mark.asyncio
    async def test_user_registration_success(async_client):
        """Test that user registration works with a random email."""
        user_data = {
            "password": "StrongPassword123!",
            "first_name": "Test",
            "last_name": "User",
            "nickname": generate_nickname(),
            "email": "john.doe@example.com",
            "role": "AUTHENTICATED"
        }
    
>       response = await async_client.post("/register/", json=user_data)
tests/test_api/test_users_api.py:77: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/httpx/_client.py:1892: in post
    return await self.request(
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/httpx/_client.py:1574: in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/httpx/_client.py:1661: in send
    response = await self._send_handling_auth(
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/httpx/_client.py:1689: in _send_handling_auth
    response = await self._send_handling_redirects(
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/httpx/_client.py:1726: in _send_handling_redirects
    response = await self._send_single_request(request)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/httpx/_client.py:1763: in _send_single_request
    response = await transport.handle_async_request(request)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/httpx/_transports/asgi.py:164: in handle_async_request
    await self.app(scope, receive, send)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/fastapi/applications.py:1054: in __call__
    await super().__call__(scope, receive, send)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/applications.py:123: in __call__
    await self.middleware_stack(scope, receive, send)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/middleware/errors.py:186: in __call__
    raise exc
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/middleware/errors.py:[164](https://github.com/so338njit/user_management/actions/runs/14939546652/job/41974198915#step:7:165): in __call__
    await self.app(scope, receive, _send)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/middleware/cors.py:83: in __call__
    await self.app(scope, receive, send)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/middleware/exceptions.py:62: in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/_exception_handler.py:64: in wrapped_app
    raise exc
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/_exception_handler.py:53: in wrapped_app
    await app(scope, receive, sender)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/routing.py:758: in __call__
    await self.middleware_stack(scope, receive, send)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/routing.py:778: in app
    await route.handle(scope, receive, send)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/routing.py:299: in handle
    await self.app(scope, receive, send)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/routing.py:79: in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/_exception_handler.py:64: in wrapped_app
    raise exc
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/_exception_handler.py:53: in wrapped_app
    await app(scope, receive, sender)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/starlette/routing.py:74: in app
    response = await func(request)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/fastapi/routing.py:278: in app
    raw_response = await run_endpoint_function(
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/site-packages/fastapi/routing.py:191: in run_endpoint_function
    return await dependant.call(**values)
app/routers/user_routes.py:209: in register_user
    user = await UserService.register_user(db, user_dict, email_service)
app/services/user_service.py:122: in register_user
    return await cls.create(session, user_data, get_email_service)
app/services/user_service.py:76: in create
    await email_service.send_verification_email(new_user)
app/services/email_service.py:33: in send_verification_email
    await self.send_user_email({
app/services/email_service.py:29: in send_user_email
    self.smtp_client.send_email(subject_map[email_type], html_content, user_data['email'])
app/utils/smtp_connection.py:25: in send_email
    server.starttls()  # Use TLS
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/smtplib.py:769: in starttls
    self.ehlo_or_helo_if_needed()
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/smtplib.py:611: in ehlo_or_helo_if_needed
    if not ([200](https://github.com/so338njit/user_management/actions/runs/14939546652/job/41974198915#step:7:201) <= self.ehlo()[0] <= 299):
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/smtplib.py:451: in ehlo
    self.putcmd(self.ehlo_msg, name or self.local_hostname)
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/smtplib.py:378: in putcmd
    self.send(f'{s}{CRLF}')
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <smtplib.SMTP object at 0x7f5bd3476540>
s = 'ehlo fv-az1277-11.rlg54ewzj0kupbrqbmaalt3b3c.ex.internal.cloudapp.net\r\n'
    def send(self, s):
        """Send `s' to the server."""
        if self.debuglevel > 0:
            self._print_debug('send:', repr(s))
        if self.sock:
            if isinstance(s, str):
                # send is used by the 'data' command, where command_encoding
                # should not be used, but 'data' needs to convert the string to
                # binary itself anyway, so that's not a problem.
                s = s.encode(self.command_encoding)
            sys.audit("smtplib.send", self, s)
            try:
                self.sock.sendall(s)
            except OSError:
                self.close()
                raise SMTPServerDisconnected('Server not connected')
        else:
>           raise SMTPServerDisconnected('please run connect() first')
E           smtplib.SMTPServerDisconnected: please run connect() first
/opt/hostedtoolcache/Python/3.12.10/x64/lib/python3.12/smtplib.py:365: SMTPServerDisconnected
------------------------------ Captured log call -------------------------------
ERROR    root:smtp_connection.py:30 Failed to send email: please run connect() first
=========================== short test summary info ============================
FAILED tests/test_api/test_users_api.py::test_user_registration_success - smtplib.SMTPServerDisconnected: please run connect() first

--- a/.github/workflows/production.yml
+++ b/.github/workflows/production.yml
@@ -50,7 +50,7 @@ jobs:
         env:
           DATABASE_URL: postgresql+asyncpg://user:password@localhost:5432/myappdb  # Configure the DATABASE_URL environment variable for tests
           SMTP_SERVER: ${{ secrets.SMTP_SERVER }}
-          SMTP_PORT: "2525"
+          SMTP_PORT: ${{ secrets.SMTP_PORT }}
           SMTP_USERNAME: ${{ secrets.SMTP_USERNAME }}
           SMTP_PASSWORD: ${{ secrets.SMTP_PASSWORD }}          
         run: pytest

--- a/app/utils/smtp_connection.py
+++ b/app/utils/smtp_connection.py
@@ -9,7 +9,7 @@ import logging
 class SMTPClient:
     def __init__(self, server: str, port: int, username: str, password: str):
         self.server = server
-        self.port = port
+        self.port = int(port) if isinstance(port, str) else port
         self.username = username
         self.password = password
 
@@ -21,11 +21,16 @@ class SMTPClient:
             message['To'] = recipient
             message.attach(MIMEText(html_content, 'html'))
 
-            with smtplib.SMTP(self.server, self.port) as server:
-                server.starttls()  # Use TLS
-                server.login(self.username, self.password)
-                server.sendmail(self.username, recipient, message.as_string())
+            # Create the SMTP connection and explicitly connect
+            server = smtplib.SMTP(self.server, self.port)
+            server.connect(self.server, self.port)  # Explicitly connect to the server
+            server.starttls()  # Use TLS
+            server.login(self.username, self.password)
+            server.sendmail(self.username, recipient, message.as_string())
+            server.quit()  # Close the connection
+            
             logging.info(f"Email sent to {recipient}")
+            return True
         except Exception as e:
             logging.error(f"Failed to send email: {str(e)}")
             raise

